### PR TITLE
feat(api): expose step stack

### LINKED ISSUE #24283: [Feature] Expose the stacks of test steps
At [Replay](https://www.replay.io/) we want to improve our support for recordings of Playwright tests. In particular we'd like to be able to show a test step's stack and its locations in the test's source, similar to Playwright's trace viewer. To do that we need Playwright to send the test step's stack to our reporter.

### PR DESCRIPTION
**Pull request by hbenl**
Fixes #24283

### CODE REVIEW

**Comment by jasonLaster** [comment]
Here's some a loom of what we're going for
https://www.loom.com/share/46192b219a9a44418262543f5f6f5733

For context, our current playwright panel is a list of locators which we feel like is difficult to map back to the test itself

File: packages/playwright-test/src/runner/dispatcher.ts [hunk 1]
```diff
@@ -267,6 +267,8 @@ export class Dispatcher {
       }
       const { result, steps } = runData;
       const parentStep = params.parentStepId ? steps.get(params.parentStepId) : undefined;
+      // Sanitize location that comes from user land, it might have extra properties.
```
Comment (pavelfeldman) [inline packages/playwright-test/src/runner/dispatcher.ts:270]: I can see where you toot it from, but I am not sure it is necessary, should be Ok to pass it in as is.

File: packages/playwright-test/src/runner/dispatcher.ts [hunk 2]
```diff
@@ -278,7 +280,8 @@ export class Dispatcher {
         startTime: new Date(params.wallTime),
         duration: -1,
         steps: [],
-        location: params.location,
+        location,
```
Comment (pavelfeldman) [inline packages/playwright-test/src/runner/dispatcher.ts:283]: You probably don't need it any longer.

File: packages/playwright-test/src/worker/testInfo.ts [hunk 3]
```diff
@@ -295,20 +295,20 @@ export class TestInfoImpl implements TestInfo {
...
       testId: this._test.id,
       stepId,
       parentStepId: parentStep ? parentStep.stepId : undefined,
       title: data.title,
       category: data.category,
       wallTime: data.wallTime,
-      location,
+      stack: data.stack?.map(frame => ({
+        ...frame,
+        file: path.relative(process.cwd(), frame.file)
```
Comment (pavelfeldman) [inline packages/playwright-test/src/worker/testInfo.ts:307]: I think we leave it up to the reporters to perform the cosmetic resolution.

File: packages/playwright-test/src/worker/testInfo.ts [hunk 4]
```diff
@@ -295,20 +295,20 @@ export class TestInfoImpl implements TestInfo {
...
+      })),
     };
     this._onStepBegin(payload);
-    this._traceEvents.push(createBeforeActionTraceEventForStep(stepId, parentStep?.stepId, data.apiName || data.title, data.params, data.wallTime, data.location ? [data.location] : []));
+    this._traceEvents.push(createBeforeActionTraceEventForStep(stepId, parentStep?.stepId, data.apiName || data.title, data.params, data.wallTime, data.stack?.length ? [data.stack[0]] : []));
```
Comment (pavelfeldman) [inline packages/playwright-test/src/worker/testInfo.ts:311]: It wants stack, you have stack, pass the stack?

File: packages/playwright-test/src/runner/dispatcher.ts (hunk 1 above)
Comment (hbenl) [reply packages/playwright-test/src/runner/dispatcher.ts:270]: Do you mean I just don't have to remove the `function` property or do you also mean I don't need to filter out a location whose `file` includes `'@playwright'`?

**Comment by pavelfeldman** [comment]
Closing to move the discussion into the issue.

**Pull request closed without merging**